### PR TITLE
#370 - Add "Assign to my team" on Projects Page

### LINKED ISSUE #370: Teams - Add "Assign to my team" on Projects Page
### Description

Create a new item on the dropdown menu for a project on the Projects page only visible to team leads called "Assign to my team" which assigns the project to their team.

### Acceptance Criteria

1. The "Assign to my team" button should only appear for team leads and should assign the project to that team lead's team without needing a change request. (calling the endpoint made in #371)
2. The button will be grayed out if the project is already assigned to a team.
This button will not appear to non-team-leads.

### Proposed Solution

Add the item to the dropdown and have it change the team that a project is assigned to (change the project's team).

### Mocks

image

### PR DESCRIPTION
**Pull request by Ari-S-123**
## Changes

Draft PR for 

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## Screenshots

_If you made UI changes you must post a screenshot of the whole page for each change in 1) a normal sized window and 2) the smallest possible window_

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

_If none of this applies, you can delete this section_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #370

### CODE REVIEW

File: src/backend/src/utils/users.utils.ts [hunk 1]
```diff
@@ -43,3 +43,14 @@ export const getUsers = async (userIds: number[]): Promise<User[]> => {
 
   return users;
 };
+
+export const getTeamForLead = (userId: number): string | undefined => {
```
Comment (anthonybernardi) [inline src/backend/src/utils/users.utils.ts:47]: unneeded

File: src/backend/src/transformers/auth-user.transformer.ts [hunk 2]
```diff
@@ -10,7 +11,8 @@ const authenticatedUserTransformer = (user: Prisma.UserGetPayload<typeof authUse
     email: user.email,
     emailId: user.emailId,
     role: user.role,
-    defaultTheme: user.userSettings?.defaultTheme
+    defaultTheme: user.userSettings?.defaultTheme,
+    teamAsLeadId: getTeamForLead(user.userId)
```
Comment (anthonybernardi) [inline src/backend/src/transformers/auth-user.transformer.ts:15]: all you need to do is change the query args to include the `teamAsLead` and then pull it out here

File: src/frontend/src/apis/projects.api.ts [hunk 3]
```diff
@@ -50,3 +50,7 @@ export const editSingleProject = (payload: any) => {
     ...payload
   });
 };
+
+export const setProjectTeam = (wbsNum: WbsNumber, team: string) => {
```
Comment (anthonybernardi) [inline src/frontend/src/apis/projects.api.ts:54]: documentation

File: src/frontend/src/hooks/projects.hooks.ts [hunk 4]
```diff
@@ -58,3 +64,19 @@ export const useEditSingleProject = (wbsNum: WbsNumber) => {
     }
   );
 };
+
+export const useSetProjectTeam = (wbsNum: WbsNumber, team: string) => {
```
Comment (anthonybernardi) [inline src/frontend/src/hooks/projects.hooks.ts:68]: documentation

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx [hunk 5]
```diff
@@ -63,6 +63,9 @@ const ProjectViewContainer: React.FC<ProjectViewContainerProps> = ({ proj, enter
       Request Change
     </MenuItem>
   );
+
+  const assignToMyTeamBtn = <MenuItem onClick={() => {}}>Assign to My Team</MenuItem>;
```
Comment (anthonybernardi) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx:67]: assignToMyTeamButton -- no need to shorten names it just makes things confusing to read

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx [hunk 6]
```diff
@@ -76,6 +79,7 @@ const ProjectViewContainer: React.FC<ProjectViewContainerProps> = ({ proj, enter
       <Menu open={dropdownOpen} anchorEl={anchorEl} onClose={handleDropdownClose}>
         {editBtn}
         {createCRBtn}
+        {auth.user?.teamAsLeadId && assignToMyTeamBtn}
```
Comment (anthonybernardi) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx:82]: if user doesnt exist you should return loading indicator earlier

**Review by anthonybernardi — CHANGES_REQUESTED** [review]

File: src/frontend/src/hooks/projects.hooks.ts [hunk 7]
```diff
@@ -58,3 +64,24 @@ export const useEditSingleProject = (wbsNum: WbsNumber) => {
...
+  const queryClient = useQueryClient();
+  return useMutation<{ message: string }, Error, any>(
+    ['projects', 'edit', 'teams'],
+    async (teamId: string) => {
+      const { data } = await setProjectTeam(wbsNum, teamId);
+      return data;
+    },
+    {
+      onSuccess: () => {
+        queryClient.invalidateQueries(['teams']);
```
Comment (anthonybernardi) [inline src/frontend/src/hooks/projects.hooks.ts:83]: this will need to invalidate `projects` too so that the projects page is updated becasue that's where we'll be doing this from anyways

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx [hunk 8]
```diff
@@ -47,12 +55,22 @@ const ProjectViewContainer: React.FC<ProjectViewContainerProps> = ({ proj, enter
     handleDropdownClose();
   };
 
-  const isGuest = auth.user?.role === 'GUEST';
+  const handleAssignToMyTeam = async () => {
+    await mutateAsync(teamAsLeadId).catch((err) => {
```
Comment (anthonybernardi) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx:59]: do a try catch it works better

File: src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx [hunk 9]
```diff
@@ -47,12 +55,22 @@ const ProjectViewContainer: React.FC<ProjectViewContainerProps> = ({ proj, enter
...
+      alert(err);
+    });
+    setProjTeamId(teamAsLeadId);
```
Comment (anthonybernardi) [inline src/frontend/src/pages/ProjectDetailPage/ProjectViewContainer/ProjectViewContainer.tsx:62]: if we invalidate the project correctly we won't need this anymore

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
nice! very close

**Review by anthonybernardi — APPROVED** [review]
awesome!
